File: internal/file/page.go
package file

import (
	"encoding/binary"
)

type Page struct {
	byteBuffer []byte
}

const int32Size = 4 // int32 is 4 bytes

func NewPage(blockSize int64) *Page {
	return &Page{
		byteBuffer: make([]byte, blockSize),
	}
}

func NewPageFromBytes(b []byte) *Page {
	return &Page{
		byteBuffer: b,
	}
}

func (p *Page) GetInt(offset int) int32 {
	return int32(binary.LittleEndian.Uint32(p.byteBuffer[offset : offset+int32Size]))
}

func (p *Page) SetInt(offset int, val int32) {
	binary.LittleEndian.PutUint32(p.byteBuffer[offset:offset+int32Size], uint32(val))
}

func (p *Page) GetBytes(offset int) []byte {
	length := p.GetInt(offset)
	return p.byteBuffer[offset+int32Size : offset+int32Size+int(length)]
}

func (p *Page) SetBytes(offset int, val []byte) {
	p.SetInt(offset, int32(len(val)))
	copy(p.byteBuffer[offset+int32Size:], val)
}

func (p *Page) GetString(offset int) string {
	return string(p.GetBytes(offset))
}

func (p *Page) SetString(offset int, val string) {
	p.SetBytes(offset, []byte(val))
}

func MaxLength(strlen int) int {
	return int32Size + strlen
}

func (p *Page) contents() []byte {
	return p.byteBuffer
}

File: app/src/main/kotlin/com/uesho/kura/file/Page.kt
package com.uesho.kura.file

import java.nio.ByteBuffer
import java.nio.charset.Charset
import java.nio.charset.StandardCharsets

/**
 * A page is a fixed-length sequence of bytes.
 * It provides methods to read and write values at any offset within the page.
 */
class Page {
    private val byteBuffer: ByteBuffer

    companion object {
        val CHARSET: Charset = StandardCharsets.US_ASCII

        /**
         * Calculate the maximum number of bytes needed to store a string of given length.
         *
         * @param strLen The length of the string
         * @return The maximum number of bytes needed
         */
        fun maxLength(strLen: Int): Int {
            val bytesPerChar = CHARSET.newEncoder().maxBytesPerChar()
            return Int.SIZE_BYTES + (strLen * bytesPerChar.toInt())
        }
    }

    /**
     * Create a new page with the specified block size.
     *
     * @param blockSize The size of the block in bytes
     */
    constructor(blockSize: Int) {
        byteBuffer = ByteBuffer.allocateDirect(blockSize)
    }

    /**
     * Create a new page from an existing byte array.
     *
     * @param bytes The byte array to create the page from
     */
    constructor(bytes: ByteArray) {
        byteBuffer = ByteBuffer.wrap(bytes)
    }

    /**
     * Get an integer value from the specified offset.
     *
     * @param offset The offset in the page
     * @return The integer value at the offset
     */
    fun getInt(offset: Int): Int = byteBuffer.getInt(offset)

    /**
     * Set an integer value at the specified offset.
     *
     * @param offset The offset in the page
     * @param value The value to set
     */
    fun setInt(offset: Int, value: Int) {
        byteBuffer.putInt(offset, value)
    }

    /**
     * Get a byte array from the specified offset.
     *
     * @param offset The offset in the page
     * @return The byte array at the offset
     */
    fun getBytes(offset: Int): ByteArray {
        byteBuffer.position(offset)
        val length = byteBuffer.getInt()
        val bytes = ByteArray(length)
        byteBuffer.get(bytes)
        return bytes
    }

    /**
     * Set a byte array at the specified offset.
     *
     * @param offset The offset in the page
     * @param bytes The byte array to set
     */
    fun setBytes(offset: Int, bytes: ByteArray) {
        byteBuffer.position(offset)
        byteBuffer.putInt(bytes.size)
        byteBuffer.put(bytes)
    }

    /**
     * Get a string from the specified offset.
     *
     * @param offset The offset in the page
     * @return The string at the offset
     */
    fun getString(offset: Int): String {
        val bytes = getBytes(offset)
        return String(bytes, CHARSET)
    }

    /**
     * Set a string at the specified offset.
     *
     * @param offset The offset in the page
     * @param value The string to set
     */
    fun setString(offset: Int, value: String) {
        val bytes = value.toByteArray(CHARSET)
        setBytes(offset, bytes)
    }

    /**
     * Get the contents of the page as a ByteBuffer.
     * This is package-private and used by FileMgr.
     */
    internal fun contents(): ByteBuffer {
        byteBuffer.position(0)
        return byteBuffer
    }
}